src/utils/resolveToolProtocol.ts

```diff
@@ -34,3 +55,47 @@ export function resolveToolProtocol(providerSettings: ProviderSettings, modelInf
 	// 3. Native Fallback
 	return TOOL_PROTOCOL.NATIVE
 }
+
+/**
+ * Detect the tool protocol used in an existing conversation history.
+ *
+ * This function scans the API conversation history for tool_use blocks
+ * and determines which protocol was used based on their structure:
+ *
+ * - Native protocol: tool_use blocks ALWAYS have an `id` field
+ * - XML protocol: tool_use blocks NEVER have an `id` field
+ *
+ * This is critical for task resumption: if a task previously used tools
+ * with a specific protocol, we must continue using that protocol even
+ * if the user's NTC settings have changed.
+ *
+ * @param messages - The API conversation history to scan
+ * @returns The detected protocol, or undefined if no tool calls were found
+ */
+export function detectToolProtocolFromHistory(messages: ApiMessageForDetection[]): ToolProtocol | undefined {
+	for (const message of messages) {
+		// Only assistant messages contain tool_use blocks
+		if (message.role !== "assistant") {
+			continue
+		}
+
+		const content = message.content
+		if (!Array.isArray(content)) {
+			continue
+		}
+
+		for (const block of content) {
+			if (block.type === "tool_use") {
+				// The presence or absence of `id` determines the protocol:
+				// - Native protocol tool calls ALWAYS have an ID (set when parsed from tool_call chunks)
+				// - XML protocol tool calls NEVER have an ID (parsed from XML text)
+				// This pattern is used in presentAssistantMessage.ts:497-500
+				const hasId = !!(block as Anthropic.ToolUseBlock).id
```

Comment:
Should we try a safer check instead of a cast? For example using the `in` operator?